Bunchkaufman- and LU-decomposition based generalized eigenvalues and eigenvectors

aravindh-krishnamoorthy opened the pull request:
# Introduction

This PR implements **Bunchkaufman (BK)-** and **LU-decomposition** based generalized eigenvalues and eigenvectors of $\boldsymbol{A}$ and $\boldsymbol{B}$ when $\boldsymbol{A}$ is arbitrary and $\boldsymbol{B}$ is (Hermitian) symmetric in the following new methods in `stdlib/LinearAlgebra/src/symmetriceigen.jl`:
1. Functions `eigvals` and `eigen`
2. In-place versions: `eigvals!` and `eigen!`

For $N \geq 32,$ these implementations **outperform the equivalent native (`eigen.jl`) functions** in both execution time and memory usage. Test and benchmark results will be [updated here regularly](https://github.com/JuliaLang/julia/pull/50471#issuecomment-1627088845).

Comments are suggestions are welcome.

- Note: This work is an overspill from PR #49673. The comments there may also be relevant for this PR.

# Open points
- [X] ~Outputs when $B$ is singular.~ (Output is analogous to the Cholesky solution.)

aravindh-krishnamoorthy committed:
Carry forward implementation from PR JuliaLang#49673

aravindh-krishnamoorthy committed:
LAPACK: Add routines for LAPMT and LAPMR, in-place column and row permutations, respectively.

aravindh-krishnamoorthy committed:
First version towards in-place

aravindh-krishnamoorthy committed:
Minor improvements to eigvals!

aravindh-krishnamoorthy committed:
First version of LAPACK based eigvals\!(A,K)

aravindh-krishnamoorthy committed:
Unify real- and complex-valued implementations.

aravindh-krishnamoorthy committed:
LAPACK based eigen\!(A, bunchkaufman(B))

aravindh-krishnamoorthy committed:
Use ordering (imag(x),real(x)) instead of (real(x),imag(x)) for improved ordering stability.

aravindh-krishnamoorthy committed:
Memory optimization: D = B.D

aravindh-krishnamoorthy committed:
Memory optimised versions

aravindh-krishnamoorthy committed:
Merge branch 'JuliaLang:master' into bk

aravindh-krishnamoorthy committed:
Code comments

aravindh-krishnamoorthy commented:
# Benchmark Results
## Benchmarking file and instructions
- The benchmarking file is here (rename to `bkmark.jl`: [bkmark.txt](https://github.com/JuliaLang/julia/files/11990417/bkmark.txt)
- Execute after checking out the PR branch:
```julia
julia> using Revise
julia> using LinearAlgebra
julia> Revise.track(LinearAlgebra)
julia> using BenchmarkTools
julia> N=32; include("/home/ark/temp/bkmark.jl")
julia> N=1000; include("/home/ark/temp/bkmark.jl")
```
## Key
**N=$N** / (**R**/**C**) / (**Native**/**BK**) / (**EV**/**E**/**EV!**/**E!**):
- **$N** is the matrix order.
- **R**: real-valued inputs, **C**: complex valued.
- **Native**: `eigen.jl`, **BK**: BK-based implementation.
- **EV**: `eigvals`, **E**: `eigen`, **EV!**: `eigvals!` and `bunchkaufman!`, **E!**: `eigen` and `bunchkaufman!`.

## Results
### $N=32$
```julia
# N=32/R/Native/EV
  184.657 μs (14 allocations: 68.59 KiB)
# N=32/R/BK/EV
  136.948 μs (26 allocations: 79.52 KiB)

# N=32/R/Native/E
  272.467 μs (16 allocations: 92.31 KiB)
# N=32/R/BK/E
  229.343 μs (32 allocations: 120.55 KiB)

# N=32/R/Native/EV!
  178.247 μs (12 allocations: 52.34 KiB)
# N=32/R/BK/EV!
  138.608 μs (21 allocations: 54.77 KiB)

# N=32/R/Native/E!
  265.489 μs (14 allocations: 76.06 KiB)
# N=32/R/BK/E!
  224.497 μs (27 allocations: 95.80 KiB)

# N=32/C/Native/EV
  564.722 μs (13 allocations: 134.53 KiB)
# N=32/C/BK/EV
  369.516 μs (26 allocations: 154.84 KiB)

# N=32/C/Native/E
  884.625 μs (15 allocations: 150.45 KiB)
# N=32/C/BK/E
  639.394 μs (36 allocations: 172.02 KiB)

# N=32/C/Native/EV!
  599.442 μs (13 allocations: 134.53 KiB)
# N=32/C/BK/EV!
  416.080 μs (21 allocations: 106.09 KiB)

# N=32/C/Native/E!
  951.423 μs (15 allocations: 150.45 KiB)
# N=32/C/BK/E!
  690.373 μs (31 allocations: 123.27 KiB)
```
### $N=1000$
```julia
# N=1000/R/Native/EV
  1.502 s (16 allocations: 16.83 MiB)
# N=1000/R/BK/EV
  381.301 ms (29 allocations: 23.75 MiB)

# N=1000/R/Native/E
  2.187 s (20 allocations: 39.70 MiB)
# N=1000/R/BK/E
  579.383 ms (38 allocations: 62.65 MiB)

# N=1000/R/Native/EV!
  1.809 s (12 allocations: 1.57 MiB)
# N=1000/R/BK/EV!
  444.229 ms (21 allocations: 869.33 KiB)

# N=1000/R/Native/E!
  2.328 s (16 allocations: 24.45 MiB)
# N=1000/R/BK/E!
  599.887 ms (30 allocations: 39.75 MiB)

# N=1000/C/Native/EV
  5.822 s (16 allocations: 33.62 MiB)
# N=1000/C/BK/EV
  1.269 s (29 allocations: 47.42 MiB)

# N=1000/C/Native/E
  7.153 s (19 allocations: 48.87 MiB)
# N=1000/C/BK/E
  1.871 s (40 allocations: 64.16 MiB)

# N=1000/C/Native/EV!
  5.750 s (16 allocations: 33.62 MiB)
# N=1000/C/BK/EV!
  1.329 s (21 allocations: 1.63 MiB)

# N=1000/C/Native/E!
  7.059 s (19 allocations: 48.87 MiB)
# N=1000/C/BK/E!
  1.858 s (32 allocations: 18.37 MiB)
```

# Test Results
## Testing file and instructions
- Tests are located in `stdlib/LinearAlgebra/test/symmetriceigen.jl`
- Execute after checking out the PR branch:
```julia
julia> using Revise
julia> using LinearAlgebra
julia> Revise.track(LinearAlgebra)
julia> include("symmetriceigen.jl")
```
## Results
```julia
Test Summary:      | Pass  Total  Time
chol-eigen-eigvals |    4      4  5.9s
Test Summary: | Pass  Total  Time
issue #49533  |   12     12  1.0s
Test Summary:   | Pass  Total  Time
bk-eigen-eigals |   12     12  2.5s
Main.TestSymmetricEigen
```

aravindh-krishnamoorthy commented:
Hello @dkarrasch and @jishnub. I look forward to your comments and suggestions on this new PR.

stevengj commented:
Is there a reference for this algorithm?

aravindh-krishnamoorthy commented:
> Is there a reference for this algorithm?

Hello @stevengj. Unfortunately, I don't have a reference handy. But, the solution follows the same method as that of the Cholesky decomposition one in [eigen/Cholesky](https://github.com/JuliaLang/julia/blob/7c805891b3afc418d6c6fb3f8f9c0d8f38e69265/stdlib/LinearAlgebra/src/symmetriceigen.jl#L180) and that in LAPACK (NIST)'s [chegv()](https://netlib.org/lapack/explore-html/d9/de3/group__complex_h_eeigen_gab2f86fb41df5ae239798c9c3081a2d49.html#gab2f86fb41df5ae239798c9c3081a2d49), and is explained below.

Consider an arbitrary matrix $\boldsymbol{A}$ and an invertible(1) Hermitian symmetric matrix $\boldsymbol{B}$ with Bunch-Kaufman decomposition given by $$\boldsymbol{B} = \boldsymbol{P}^{\mathrm{H}}\boldsymbol{U}\boldsymbol{D}\boldsymbol{U}^{\mathrm{H}}\boldsymbol{P},$$ where $\boldsymbol{P}$ is a permutation matrix, $\boldsymbol{U}$ is an invertible upper (lower) triangular matrix, and $\boldsymbol{D}$ is a Hermitian symmetric $2\times 2$-block diagonal matrix (a special case of tridiagonal), as given in [LinearAlgebra.bunchkaufman](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.bunchkaufman).

### Eigenvalues
We note that the roots of the generalised eigenvalue problem $$\det(\boldsymbol{A} - \lambda\boldsymbol{B})$$ and those of $$\det\big(\boldsymbol{D}^{-1}\boldsymbol{U}^{-1}(\boldsymbol{P}\boldsymbol{A}\boldsymbol{P}^{\mathrm{H}})(\boldsymbol{U}^{\mathrm{H}})^{-1} - \lambda\boldsymbol{I}\big)$$ are equivalent. Hence, the generalised eigenvalues of $\boldsymbol{A}$ and $\boldsymbol{B}$ are the eigenvalues of $\boldsymbol{A}' = \boldsymbol{D}^{-1}\boldsymbol{U}^{-1}(\boldsymbol{P}\boldsymbol{A}\boldsymbol{P}^{\mathrm{H}})(\boldsymbol{U}^{\mathrm{H}})^{-1}.$ Let $\boldsymbol{\Lambda}$ be a matrix with the eigenvalues of $\boldsymbol{A}'$ on the main diagonal.

### Eigenvectors
Now, let $\boldsymbol{V}'$ contain the eigenvectors of $\boldsymbol{A}'$ in its columns. Then, the generalised eigenvectors, based on the above, are contained in the columns of $$\boldsymbol{V} = \boldsymbol{P}^{\mathrm{H}}(\boldsymbol{U}^{\mathrm{H}})^{-1}\boldsymbol{V}'.$$

These satisfy the relation $$\boldsymbol{A}\boldsymbol{V}  = \boldsymbol{B}\boldsymbol{V}\boldsymbol{\Lambda},$$ which is utilized for testing the functions.

### Code
In the proposed code, these above performed in-place and step-by-step using LAPACK functions. The code comments indicate which step is being performed in the following line. Please let me know in case you have questions or concerns about numerical stability of the approach. I feel that, in principle, these should not be a problem with LAPACK functions and Bunch-Kaufman decomposition, which is quite stable like the Cholesky decomposition.

Comments and suggestions are welcome.

(1) For non-invertible $\boldsymbol{B},$ the usual convention is to mark the corresponding eigenvalue with infinity. Julia returns `NaN+Nan*im`. This algorithm returns numbers of the order $10^{16}$ in place of infinity. A discussion on this is still open, please see checkbox 3 in the PR text.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,50 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
 end
 
+# Bunch-Kaufmann (LDLT) based solution for generalized eigenvalues and eigenvectors
+function eigen(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
```
I suspect you need 
```suggestion
function eigen(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:BlasFloat}
```
here, don't you? And similarly below.

aravindh-krishnamoorthy replied:
Thank you for this comment. I agree. Upon noting that `BlasReal <: BlasFloat` and `BlasComplex <: BlasFloat`, I've changed `Number` to `BlasFloat` in all the four relevant method definitions. The change is in https://github.com/aravindh-krishnamoorthy/julia/commit/8b8f556d7095452e14a3bc7c46c4bc2f34d6bb9d.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,50 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
 
+# Bunch-Kaufmann (LDLT) based solution for generalized eigenvalues and eigenvectors
+function eigen(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    eigen!(copy(A), copy(B); sortby)
```
Is your algorithm mutating `B`? If not, then you don't need the copy, and the added function in `bunchkaufman.jl`.

aravindh-krishnamoorthy replied:
Unfortunately, a call to `syconv!` or `syconvf_rook!` is in-place and modifies `B.LD`. The usual `getproperty(B, :L)` creates a copy. Hence, the copy `copy(B)` is necessary.

dkarrasch replied:
Okay, but if both calls make a copy, then why not leave it to `getproperty`?

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,50 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+function eigen(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    eigen!(copy(A), copy(B); sortby)
+end
+function eigen!(A::AbstractMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
```
I believe that whenever we have direct calls to `LAPACK`, we restrict the argument type to `StridedMatrix`, most likely together with eltype restrictions like `BlasFloat` or `BlasReal` or something.

aravindh-krishnamoorthy replied:
Thank you for the comment, I agree. I've incorporated both the `StridedMatrix` and `BlasFloat` changes in the following commit: https://github.com/aravindh-krishnamoorthy/julia/commit/8b8f556d7095452e14a3bc7c46c4bc2f34d6bb9d.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,50 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+        M = UnitUpperTriangular(LUD)
+        du = od[2:end]
+        # Aliasing of dl and du is not allowed for gtsv! below.
+        dl = B.symmetric ? copy(du) : conj.(copy(du))
```
Applying broadcasting to a vector, even if it's `identity`, yields an unaliased result.
```suggestion
        dl = B.symmetric ? copy(du) : conj.(du)
```
Similarly below.

aravindh-krishnamoorthy replied:
Thank you for this comment and the information. As you're aware, I've only started using Julia seriously since a few months, and information such as this is very helpful. I've made this change as well as a part of https://github.com/aravindh-krishnamoorthy/julia/commit/8b8f556d7095452e14a3bc7c46c4bc2f34d6bb9d.

dkarrasch commented on stdlib/LinearAlgebra/test/symmetriceigen.jl:
```diff
@@ -8,7 +8,7 @@ using Test, LinearAlgebra
 
     # eigenvalue sorting
-    sf = x->(real(x),imag(x))
+    sf = x->(imag(x),real(x))
```
Why changing the sorting function in unrelated tests?

aravindh-krishnamoorthy replied:
Thank you for this comment. Indeed, this change looks weird, but is unfortunately necessary. I discuss it in the post below.

dkarrasch reviewed (commented):
Thanks for tackling this. I see this is addressing one of my major concerns, that the algorithm is not really in-place of the left-hand matrix! Cool. Here are a couple of first comments.

aravindh-krishnamoorthy committed:
Incorporate first round of comments from @dkarrasch

aravindh-krishnamoorthy commented:
> Thanks for tackling this. I see this is addressing one of my major concerns, that the algorithm is not really in-place of the left-hand matrix! Cool. Here are a couple of first comments.

Hello @dkarrasch. Thank you for the first round of comments. I've responded to the comments individually and incorporated the changes via a single commit https://github.com/aravindh-krishnamoorthy/julia/commit/8b8f556d7095452e14a3bc7c46c4bc2f34d6bb9d.

Below, I'd like to address the following comment.
> Why changing the sorting function in unrelated tests?

In the tests in `test/symmetriceigen.jl`, besides the test $\boldsymbol{A}\boldsymbol{V}  = \boldsymbol{B}\boldsymbol{V}\boldsymbol{\Lambda},$ the outputs of `dense.jl`'s generalised eigenvalues and those from BK are compared. Gaussian random matrices are used as test matrices(1). Hence, the test matrices are almost surely full-rank and have eigenvalue separation (magnitude) $> \epsilon.$ Furthermore, most eigenvalues are in pairs $(a \pm \mathrm{i}b).$

However, due to numerical issues, the eigenvalue pairs between the functions differ as $(a + \mathrm{i}b)$ and $(a + \epsilon - \mathrm{i}b),$ which reverses the ordering causing the comparisons to fail. Arranging them by `x->(imag(x),real(x))` solves this issue.

(1) However, even though using Gaussian random matrices as test matrices is advantageous (greater coverage as the matrices vary every test), it is still possible that we end up with rank deficient or closely spaced eigenvalues. In such cases, the resulting test failures will be difficult to reproduce. If you prefer to not have this uncertainty, I can locally check with Gaussian random matrices but have a fixed matrix in the official test.

Kindly let me know your opinion.

dkarrasch commented:
I think this is a great opportunity to clean-up/extend a few things. I'd suggest to use more generic functions to achieve the same thing, but for specific cases use LAPACK functions. For instance:

* regarding the row and column permutation, we already have https://github.com/JuliaLang/julia/blob/8dd58a48397d7a43e659e3e79fd7027107dbba15/stdlib/LinearAlgebra/src/lu.jl#L373-L418 My suspicion is that it works the same for both `LU` and `BunchKaufman` (to be checked). In that case, one could simply extend the method signature. The question is whether LAPACK's version is any faster, in which case we might redirect the case of `BlasFloat`-`StridedMatrix`es to your added functions. If not, then we may not need to add the LAPACK functions.
* regarding the definition of `M`, `du` etc.: This looks like code duplication to `getproperty` of `BunchKaufman`. Then why not simply define `D = B.D` and `M = B.uplo == 'U' ? B.U : B.L`? If the issue is later that we need un-aliased off-diagonals, then I think we should change that in `getproperty` and make it un-aliased in all cases, and not sometimes yes and sometimes not. The extra effort is relatively small.
* The next lines have well-established generic function calls:
    * "# In-place A .= inv(M)*A" is `ldiv!(M, A)`
    * "# In-place A .= A*inv(M^H)" is `rdiv!(A, M')`
    * "# In-place A .= inv(Tridiagonal(dl, d = diag(LUD), du))*A" could be `ldiv!(lu!(D), A)`.
    * "vecs .= inv(M^H)*vecs" is again just `ldiv!(M', vecs)`
    * "# In-place vecs .= vecs[invperm(B.p), :]" might be done with one of the `apply_...` functions referenced above.

aravindh-krishnamoorthy committed:
Add generic implementations.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,59 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    vecs .= ldiv!(M', vecs)[invperm(B.p), :]
+    GeneralizedEigen(sorteig!(vals, vecs, sortby)...)
+end
+function eigen!(A::StridedMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:BlasFloat}
```
Is this method really needed? The `[l/r]div!` etc. above should use those LAPACK functions via multiple dispatch.

aravindh-krishnamoorthy replied:
Thank you for this comment. Please see my post below for my response.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,59 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+function eigen!(A::AbstractMatrix{T}, B::BunchKaufman{T,<:AbstractMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:Number}
+    D = copy(B.D)
+    M = B.uplo == 'U' ? B.U : B.L
+    A .= A[B.p, B.p]
```
This way, the rhs will first be allocated, and then copied to `A`. Can't we use the `apply_*` functions that I referenced, in order to do it completely allocation-free?

aravindh-krishnamoorthy replied:
Thank you for this comment. Please see my post below for my response.

aravindh-krishnamoorthy committed:
Updates to eigen! and eigvals! for BK -- align from LAPACK to Julia routines.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,45 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+end
+function eigen!(A::StridedMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:BlasFloat}
+   # NOTE: Copy of BunchKaufman's getproperty(B, :D) and getproperty(B, :L/U) in bunchkaufman.jl as 'getproperty' is not in place.
+   if B.rook
+        LUD, od = LAPACK.syconvf_rook!(B.uplo, 'C', B.LD, B.ipiv)
+    else
+        LUD, od = LAPACK.syconv!(B.uplo, B.LD, B.ipiv)
+    end
+    if B.uplo == 'U'
+        M = UnitUpperTriangular(LUD)
+        du = od[2:end]
+        # Aliasing of dl and du is not allowed for lu!(Tridiagonal(dl, diag(LUD), du) below.
+        dl = B.symmetric ? copy(du) : conj.(du)
+    else
+        M = UnitLowerTriangular(LUD)
+        dl = od[1:end-1]
+        # Aliasing of dl and du is not allowed for lu!(Tridiagonal(dl, diag(LUD), du) below.
+        du = B.symmetric ? copy(dl) : conj.(dl)
+    end
```
Why can't this just be

```suggestion
   D = B.D
   M = B.uplo == 'U' ? B.U : B.L
```
? The code above is also not in-place; in fact, it does exactly the same as `getproperty`, so should be the same effort with much less code?

aravindh-krishnamoorthy replied:
Please see https://github.com/JuliaLang/julia/pull/50471#issuecomment-1636889413

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,45 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    # Compute generalized eigenvalues of equivalent matrix:
+    #    A' = inv(Tridiagonal(dl,d,du))*inv(M)*P*A*P'*inv(M')
+    # See: https://github.com/JuliaLang/julia/pull/50471#issuecomment-1627836781
+    LAPACK.lapmt!(A, B.p, true)
+    LAPACK.lapmr!(A, B.p, true)
```
Is there a performance difference between these functions and the `apply_*_permutation` things?

aravindh-krishnamoorthy replied:
Thank you for this comment. Please see https://github.com/JuliaLang/julia/pull/50471#issuecomment-1636889413.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,45 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    LAPACK.lapmr!(A, B.p, true)
+    ldiv!(M, A)
+    rdiv!(A, M')
+    ldiv!(lu!(Tridiagonal(dl, diag(LUD), du)), A)
```
I think we should just make sure that `B.D` returns a `Tridiagonal` with unaliased diagonals, and then we might be able to use https://github.com/JuliaLang/julia/pull/50535, i.e, `ldiv!` directly on `B.D`.

aravindh-krishnamoorthy replied:
Thank you for this comment. Please see https://github.com/JuliaLang/julia/pull/50471#issuecomment-1636889413.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,45 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    # Compute generalized eigenvectors from 'vecs':
+    #   vecs = P'*inv(M')*vecs
+    # See: https://github.com/JuliaLang/julia/pull/50471#issuecomment-1627836781
+    LAPACK.trtrs!(B.uplo, 'C', 'U', convert(typeof(vecs), LUD), vecs)
```
Isn't this just

```suggestion
    ldiv!(M', vecs)
```
?

aravindh-krishnamoorthy replied:
Incorporated!

dkarrasch reviewed (commented):
A few more suggestions for simplification and genericity. It would be good to use only generic functions (i.e., not use any LAPACK functions directly), and then not restrict on `BlasFloat` unnecessarily.

aravindh-krishnamoorthy committed:
Updates to eigen! for BK -- align from LAPACK to Julia routines.

aravindh-krishnamoorthy commented:
Hello @dkarrasch. Thank you very much for your support with this PR and your comments. Sorry for the delay in my response.

Please note that the intermediate version was inadvertently pushed to GitHub. It was primarily for my benchmarking use. I've removed the function and reverted the signatures.

> I think this is a great opportunity to clean-up/extend a few things. I'd suggest to use more generic functions to achieve the same thing, but for specific cases use LAPACK functions. 

Thank you. Indeed, my idea is to first understand the most optimal performance via BLAS/LAPACK and then ensure that the final Julia code can get there. I've incorporated most of the suggestions below. Unfortunately, in a few cases, the LAPACK routines were superior either in terms of complexity or memory usage. Any suggestions to rewrite them in Julia will be welcome. Details are as follows.

> For instance:
> 
> * regarding the row and column permutation, we already have https://github.com/JuliaLang/julia/blob/8dd58a48397d7a43e659e3e79fd7027107dbba15/stdlib/LinearAlgebra/src/lu.jl#L373-L418
>    My suspicion is that it works the same for both `LU` and `BunchKaufman` (to be checked). In that case, one could simply extend the method signature. The question is whether LAPACK's version is any faster, in which case we might redirect the case of `BlasFloat`-`StridedMatrix`es to your added functions. If not, then we may not need to add the LAPACK functions.

Unfortunately, the routines are for the LAPACK-style swap permutations and are best suited for the IPIV-style permutation vectors. Using these methods for permuting $A$ will entail additional complexity and memory.

For permuting $A,$ I feel that the LAPACK routines are the superior alternative. If necessary, these LAPACK routines can also be wrapped in a `permute! :: StridedMatrix -> Vector -> StridedMatrix` method to make them look more Julia-like.

In our case, indeed, obtaining `B.L/U` from `B.LD` in BK necessitates an IPIV permutation. However, presently, BK is only supported via LAPACK and, in this case, this job is best done by `syconvf/_rook!`, which additionally also extract the super/sub diagonals.

Hence, these routines may not be beneficial for our case.

> * regarding the definition of `M`, `du` etc.: This looks like code duplication to `getproperty` of `BunchKaufman`. Then why not simply define `D = B.D` and `M = B.uplo == 'U' ? B.U : B.L`? If the issue is later that we need un-aliased off-diagonals, then I think we should change that in `getproperty` and make it un-aliased in all cases, and not sometimes yes and sometimes not. The extra effort is relatively small.

Unfortunately, the `getproperty` code of BK is inefficient. It calls `syconvf/_rook!` over-and-over again for each invocation, and additionally, utilizes memory, as seen below.

```julia
julia> B = complex.(randn(1000,1000),randn(1000,1000)) ; B = (B+B')/2 ;
julia> K = bunchkaufman(B) ;
julia> @time K.D ;
  0.002521 seconds (7 allocations: 15.320 MiB)
julia> @time K.uplo == 'U' ? K.U : K.L ;
  0.010232 seconds (5 allocations: 15.274 MiB)
```

On the other hand, as seen from the benchmark in the first comment above, the present code with duplicated `getproperty` is almost completely in place and requires less memory than even the generic counterpart.

> * The next lines have well-established generic function calls:
>   
>   * "# In-place A .= inv(M)*A" is `ldiv!(M, A)`
>   * "# In-place A .= A*inv(M^H)" is `rdiv!(A, M')`
>   * "# In-place A .= inv(Tridiagonal(dl, d = diag(LUD), du))*A" could be `ldiv!(lu!(D), A)`.
>   * "vecs .= inv(M^H)*vecs" is again just `ldiv!(M', vecs)`

The above four suggestions are incorporated!

>   * "# In-place vecs .= vecs[invperm(B.p), :]" might be done with one of the `apply_...` functions referenced above.

Unfortunately, as explained earlier, this may not be possible.

dkarrasch commented on stdlib/LinearAlgebra/src/lapack.jl:
```diff
@@ -6941,4 +6941,48 @@ Returns `X` (overwriting `C`) and `scale`.
+            m, n = size(A)
+            lda = max(1, stride(A, 2))
+            lp = length(p)
+            if n != lp
+                throw(DimensionMismatch("The second dimension of A, ($m,$n), and P, ($lp), must match"))
+            end
```
```suggestion
            if m != lp
                throw(DimensionMismatch("The first dimension of A, ($m,$n), and P, ($lp), must match"))
            end
```

dkarrasch reviewed (commented):
One quick suggestion, but actually, I think we should implement these in julia. After all, this is just a helper function. I'll post a suggestion shortly.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,46 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    # See: https://github.com/JuliaLang/julia/pull/50471#issuecomment-1627836781
+    M = B.uplo == 'U' ? UnitUpperTriangular{eltype(vecs)}(M) : UnitLowerTriangular{eltype(vecs)}(M) ;
+    ldiv!(M', vecs)
+    LAPACK.lapmr!(vecs, invperm(B.p), true)
```
Isn't this

```suggestion
    LAPACK.lapmr!(vecs, B.p, false)
```
?

aravindh-krishnamoorthy committed:
Apply suggestions from code review @dkarrasch

Co-authored-by: Daniel Karrasch <daniel.karrasch@posteo.de>

aravindh-krishnamoorthy commented:
Hello @dkarrasch, please just wait with this. I want to compare this approach to an analogous LU-decomposition based approach and ensure that BK is the right way to go.

aravindh-krishnamoorthy committed:
Implement eigen and eigvals for LU decomposition.

aravindh-krishnamoorthy commented:
Hello @dkarrasch, I now have the LU-decomposition version (for square matrices, which include symmetric and Hermitian matrices) as well. **The performance of LU-decomposition methods is close to that of the BK-decomposition methods.** But, the memory consumption is a bit high for the in-place functions. This is because, unlike BK, `getproperty` function of LU-decomposition (which allocates memory for each call) cannot be incorporated into the `eig/en/vals` functions.

I've updated the test cases, test results and the benchmark results in the comments above.

I feel it may be better to keep both the LU and BK versions. However, I look forward to your views on this.

dkarrasch commented:
Hi! I'd like to propose to add the following methods for column and row permutations in `combinatorics.jl` in `Base`:

```julia
permutecols!(a::AbstractMatrix, p::AbstractVector{<:Integer}) =
    _permute!(a, p, Base.swapcols!)
permuterows!(a::AbstractMatrix, p::AbstractVector{<:Integer}) =
    _permute!(a, p, Base.swaprows!)
function _permute!(a::AbstractMatrix, p::AbstractVector{<:Integer}, swapfun!::F) where {F}
    require_one_based_indexing(a, p)
    p .= .-p
    for i in 1:length(p)
        p[i] > 0 && continue
        j = i
        in = p[j] = -p[j]
        while p[in] < 0
            swapfun!(a, in, j)
            j = in
            in = p[in] = -p[in]
        end
    end
    a
end
invpermutecols!(a::AbstractMatrix, p::AbstractVector{<:Integer}) =
    _invpermute!(a, p, Base.swapcols!)
invpermuterows!(a::AbstractMatrix, p::AbstractVector{<:Integer}) =
    _invpermute!(a, p, Base.swaprows!)
function _invpermute!(a::AbstractMatrix, p::AbstractVector{<:Integer}, swapfun!::F) where {F}
    require_one_based_indexing(a, p)
    p .= .-p
    for i in 1:length(p)
        p[i] > 0 && continue
        j = p[i] = -p[i]
        while j != i
           swapfun!(a, j, i)
           j = p[j] = -p[j]
        end
     end
    a
end
```

They are straightforward translations of the LAPACK functions, and the row permutation (IIRC) is even faster than LAPACK. They are written such that `p` doesn't get destroyed in passing, so can be reused, in contrast to the currently used `permutecols!!`. We may decide to remove it or perhaps to redirect it to `permutecols!`.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,65 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    LAPACK.lapmr!(A, B.p, true)
+    ldiv!(M, A)
+    rdiv!(A, M')
+    ldiv!(lu!(Tridiagonal(dl, diag(LUD), du)), A)
```
We now have a (destructive) `ldiv!` for `Tridiagonal`s.

```suggestion
    ldiv!(Tridiagonal(dl, diag(LUD), du), A)
```

IIUC, the diagonals do not alias with the original data in `B`, so even their destruction in `ldiv!` does not modify the `BunchKaufman`.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,65 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
 
+# Bunch-Kaufmann (LDLT) based solution for generalized eigenvalues and eigenvectors
+function eigen(A::StridedMatrix{T}, B::BunchKaufman{T,<:AbstractMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:BlasFloat}
+    eigen!(copy(A), copy(B); sortby)
```
Does this call modify the second argument? It doesn't seem so, but we should check. Like, remove the copy here and then see if `B` is the same after the whole `eigen` computation.

aravindh-krishnamoorthy replied:
Unfortunately, the line `M, TD, p = getproperties!(B)` modifies `B`.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,65 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    # Compute generalized eigenvalues of equivalent matrix:
+    #    A' = inv(Tridiagonal(dl,d,du))*inv(M)*P*A*P'*inv(M')
+    # See: https://github.com/JuliaLang/julia/pull/50471#issuecomment-1627836781
+    LAPACK.lapmt!(A, B.p, true)
+    LAPACK.lapmr!(A, B.p, true)
```
These two could be replaced by the generic permutation functions.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,65 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    # See: https://github.com/JuliaLang/julia/pull/50471#issuecomment-1627836781
+    M = B.uplo == 'U' ? UnitUpperTriangular{eltype(vecs)}(M) : UnitLowerTriangular{eltype(vecs)}(M) ;
+    ldiv!(M', vecs)
+    LAPACK.lapmr!(vecs, B.p, false)
```
Same here with the generic functions.

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,65 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+function eigen!(A::StridedMatrix{T}, F::LU{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T}
+    L = UnitLowerTriangular(F.L)
+    U = UpperTriangular(F.U)
+    LAPACK.lapmr!(A, F.p, true)
```
See above.

dkarrasch reviewed (commented):
A few more comments. Moreover, there's a generic `BunchKaufman` decomposition in the pipeline (#51487), so we should keep this as generic as possible w.r.t. the eltype. After all, this is a generic linear algebra algorithm, not a LAPACK specific one.

aravindh-krishnamoorthy committed:
Update stdlib/LinearAlgebra/src/symmetriceigen.jl

Co-authored-by: Daniel Karrasch <daniel.karrasch@posteo.de>

aravindh-krishnamoorthy committed:
Incorporate row and column permutations for AbstractMatrices in Base/combinatorics.jl, as proposed by @dkarrasch

Co-authored-by: Daniel Karrasch <daniel.karrasch@posteo.de>

aravindh-krishnamoorthy committed:
Merge branch 'master' into bk

aravindh-krishnamoorthy committed:
Minor spelling correction to symmetriceigen test suite.

aravindh-krishnamoorthy committed:
Replace LAPACK's lapmr! and lapmt! by Julia Base's (proposed) permuterows!, permutecols!, and invpermuterows!

dkarrasch commented on stdlib/LinearAlgebra/src/symmetriceigen.jl:
```diff
@@ -184,6 +184,65 @@ function eigen!(A::AbstractMatrix, C::Cholesky; sortby::Union{Function,Nothing}=
+    eigen!(copy(A), copy(B); sortby)
+end
+function eigen!(A::StridedMatrix{T}, B::BunchKaufman{T,<:StridedMatrix}; sortby::Union{Function,Nothing}=nothing) where {T<:BlasFloat}
+   # NOTE: Copy of BunchKaufman's getproperty(B, :D) and getproperty(B, :L/U) in bunchkaufman.jl as 'getproperty' is not in place.
+   if B.rook
+        LUD, od = LAPACK.syconvf_rook!(B.uplo, 'C', B.LD, B.ipiv)
+    else
+        LUD, od = LAPACK.syconv!(B.uplo, B.LD, B.ipiv)
+    end
+    if B.uplo == 'U'
+        M = UnitUpperTriangular(LUD)
+        du = od[2:end]
+        # Aliasing of dl and du is not allowed for lu!(Tridiagonal(dl, diag(LUD), du) below.
+        dl = B.symmetric ? copy(du) : conj.(du)
+    else
+        M = UnitLowerTriangular(LUD)
+        dl = od[1:end-1]
+        # Aliasing of dl and du is not allowed for lu!(Tridiagonal(dl, diag(LUD), du) below.
+        du = B.symmetric ? copy(dl) : conj.(dl)
+    end
```
How about moving this to `bunchkaufman.jl`, and call it `_getproperties`, or `bk` inserted somewhere? Perhaps it's gonna be a slightly type-unstable function, but so is this code in here. Also, it could be reused then in cases where all properties of `BunchKaufman` are needed.

aravindh-krishnamoorthy replied:
Thank you for making the code generic! In fact, the above is a sore spot that I wanted to also discuss with you. The current `B.D` and `B.L` call the `getproperty` function twice. The above function returns both `B.D` and `B.L` in a single call, and also reuses the memory in `LUD`. Is there any way we can do this neatly?

dkarrasch replied:
That's what I mean: make a function `getproperties(B::BunchKaufman) -> (B.D, B.L, B.p)` with exactly the code that you have, i.e., call `syconv*!` only once and return "all" useful properties at once: the `Tridiagonal` (with unaliased off-diagonals), the `Unit*Triangular`, and the permutation. As it stands, you have this code now twice, which calls for hoisting it out. I imagine that this is useful to any user who wants to write performance-optimized code involving `BunchKaufman`. We should then document that function and announce it in the NEWS, as we should announce the general new feature here, actually.

aravindh-krishnamoorthy replied:
Perfect! Sounds good...

dkarrasch replied:
The method signature should just be tight enough that it catches the LAPACK cases and leaves room for another, generic method in that other BK-PR (not sure though if B.L, B.D "suffers" from the same double function call issue, but if yes, this would be even more helpful).

dkarrasch commented on base/combinatorics.jl:
```diff
@@ -267,6 +267,44 @@ end
 
 invperm(P::Any32) = Tuple(invperm(collect(P)))
 
+# Row and column permutations for AbstractMatrix
```
Should this code be moved closer to `permutecols!!` in line 119ff?

aravindh-krishnamoorthy replied:
Done!

dkarrasch reviewed (commented):
What a beautifully generic piece of code!!! If we keep the LAPACK permutation functions, then they need to be tested. Since they are new and still unused internally, perhaps we could leave them out here?

aravindh-krishnamoorthy committed:
Implement in-place function  getproperties! in bunchkaufmann.jl

aravindh-krishnamoorthy commented:
Thank you, @dkarrasch. Now, the code indeed is beautiful and generic. The main changes are done. I will review and benchmark everything again (esp with the new permutation routines) and incorporate other cosmetic changes soon.

aravindh-krishnamoorthy committed:
Move permuterows!, permutecols!, and invpermuterows!
close to permutecols!! in Julia Bases combinatorics.jl

aravindh-krishnamoorthy commented:
@dkarrasch Perhaps it makes sense to wait for #51763 to be merged so that `getproperties!` in `bunchkaufman.jl` can be cleaned up in this PR a bit?

aravindh-krishnamoorthy commented:
Hello @dkarrasch. I see that #51763 is merged. Now, I'll resume work on this one incorporating those changes. Hope to finalise this PR soon.

aravindh-krishnamoorthy committed:
Merge branch 'master' into bk

aravindh-krishnamoorthy committed:
Update the bunchkaufman.jl/getproperties! function to align to the anti-aliasing implementation in #51763 + include comments in symmetriceigen.jl

dkarrasch commented on stdlib/LinearAlgebra/src/lapack.jl:
```diff
@@ -7045,4 +7045,48 @@ julia> LAPACK.lacpy!(B, A, 'U')
+                (:zlapmt_, :ComplexF64))
+    @eval begin
+        function lapmt!(A::AbstractMatrix{$elty}, p::AbstractVector{<:BlasInt}, fb::Bool)
+            require_one_based_indexing(A)
+            chkstride1(A)
+            m, n = size(A)
+            lda = max(1, stride(A, 2))
+            lp = length(p)
+            if n != lp
+                throw(DimensionMismatch("The second dimension of A, ($m,$n), and P, ($lp), must match"))
+            end
+            ccall((@blasfunc($fn), libblastrampoline), Cvoid,
+            (Ref{BlasInt}, Ref{BlasInt}, Ref{BlasInt}, Ptr{Float64}, Ref{BlasInt}, Ptr{BlasInt}),
+            fb, m, n, A, lda, p) ;
+        end
+    end
+end
+
+# Routines for row permutation
+for (fn, elty) in ((:slapmr_, :Float32),
+                (:dlapmr_, :Float64),
+                (:clapmr_, :ComplexF32),
+                (:zlapmr_, :ComplexF64))
+    @eval begin
+        function lapmr!(A::AbstractMatrix{$elty}, p::AbstractVector{<:BlasInt}, fb::Bool)
+            require_one_based_indexing(A)
+            chkstride1(A)
+            m, n = size(A)
+            lda = max(1, stride(A, 2))
+            lp = length(p)
+            if m != lp
+                throw(DimensionMismatch("The first dimension of A, ($m,$n), and P, ($lp), must match"))
+            end
+            ccall((@blasfunc($fn), libblastrampoline), Cvoid,
+            (Ref{BlasInt}, Ref{BlasInt}, Ref{BlasInt}, Ptr{Float64}, Ref{BlasInt}, Ptr{BlasInt}),
+            fb, m, n, A, lda, p) ;
+        end
+    end
+end
+
```
Do we still need these? Given the generic ones we have, having these adds testing and maintenance burden.

aravindh-krishnamoorthy replied:
Thank you very much, @dkarrasch. I'm truly indebted to your support and guidance. Indeed, these can go, and I will remove them soon. Soon after, I'll rerun the benchmark tests, analyse a few things, update NEWS.md, and let you know as soon as things are finalised!

dkarrasch reviewed (commented):
Last "hygiene" comment, and then this is finally ready to go. Please add an entry in NEWS.md that announces these features.

aravindh-krishnamoorthy committed:
Remove unused permutation routines lapmt! and lapmr! from lapack.jl

aravindh-krishnamoorthy committed:
BUGFIX: Remove unused permutation routines lapmt! and lapmr! from lapack.jl

aravindh-krishnamoorthy committed:
Update NEWS.md

aravindh-krishnamoorthy commented:
@dkarrasch This now looks good to go from my side. I looked at the benchmarks and they seem Ok now. For both cases (N=32, 1000), the new functions significantly outperform the generic function (not optimised for symmetric matrices). Furthermore, I feel it makes sense to retain both Bunchkaufman and LU functions since each is optimal in a certain scenario. Also, I've completed a few last-minute checks and updated NEWS.md.

Nevertheless, I do welcome any last-minute comments from you and will be happy to incorporate them -- ark

aravindh-krishnamoorthy committed:
Remove redundant and wrong definition of sf in test/symmetriceigen.jl

dkarrasch commented:
To be "fair", one should perhaps include the factorization of `B` in the measurement. Did you do that? In any case, we're not changing any defaults here, so users can benchmark their use cases and make a call. Having the factorization available may be helpful or even necessary in some cases, depending on the context.

aravindh-krishnamoorthy commented:
> To be "fair", one should perhaps include the factorization of `B` in the measurement. Did you do that? In any case, we're not changing any defaults here, so users can benchmark their use cases and make a call. Having the factorization available may be helpful or even necessary in some cases, depending on the context.

Indeed, the factorisation overheads due to `bunchkaufman` and `lu` are included in the runtime and memory benchmark figures given above.
